srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,957 @@
+		#[weight = SimpleDispatchInfo::FixedNormal(750_000)]
+		fn vote(origin, votes: Vec<T::AccountId>) {
+			let who = ensure_signed(origin)?;
+
+			// TODO: use decode_len #3071.
+			let candidates = Self::candidates();
+			ensure!(!candidates.is_empty(), "number of candidates cannot be zero");
+			ensure!(candidates.len() >= votes.len(), "cannot vote more than candidates");
+
+			ensure!(!votes.is_empty(), "must vote for at least one candidate.");
+			ensure!(!Self::is_old_voter(&who), "must first remove self as voter and re-submit");
+
+			if !Self::is_current_voter(&who) {
+				// Amount to be locked up.
+				let locked_balance = T::Currency::total_balance(&who);
+
+				// lock and reserve
+				T::Currency::reserve(&who, T::VotingBond::get())
+					.map_err(|_| "voter can not pay voting bond")?;
+				T::Currency::set_lock(
+					MODULE_ID,
+					&who,
+					locked_balance,
+					T::BlockNumber::max_value(),
+					WithdrawReasons::except(WithdrawReason::TransactionPayment),
+				);
+				<StakeOf<T>>::insert(&who, locked_balance);
+			}
+
+			let now = Self::election_rounds();
+			<VotesOf<T>>::insert(&who, (votes, now));
+		}
+
+		/// Remove `origin` as a voter. One can use this to
+		///   1. _undo_ their votes before the election has started.
+		///   2. Remove their lock and unreserve the bond after an election round.
+		#[weight = SimpleDispatchInfo::FixedNormal(500_000)]
+		fn remove_voter(origin) {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				Self::is_current_voter(&who) || <StakeOf<T>>::exists(&who),
+				"should be a current/old voter"
+			);
+
+			// remove storage.
+			<VotesOf<T>>::remove(&who);
+			<StakeOf<T>>::remove(&who);
+
+			// unreserve the bond and remove.
+			T::Currency::unreserve(&who, T::VotingBond::get());
+			T::Currency::remove_lock(MODULE_ID, &who);
+		}
+
+		/// Submit oneself for candidacy.
+		///
+		/// A candidate will either:
+		///   - Lose at the end of the term and gets their bond unreserved
+		///   - Win and become a member. Members will get their bond back at the end ot their term.
+		#[weight = SimpleDispatchInfo::FixedNormal(1_000_000)]
+		fn submit_candidacy(origin) {
+			let who = ensure_signed(origin)?;
+
+			let is_candidate = Self::is_candidate(&who);
+			ensure!(!is_candidate.is_ok(), "duplicate candidate submission");
+			// assured to be an error, error always contains the index.
+			let index = is_candidate.unwrap_err();
+
+			T::Currency::reserve(&who, T::CandidacyBond::get())
+				.map_err(|_| "candidate has not enough funds")?;
+
+			<Candidates<T>>::mutate(|c| c.insert(index, who));
+			CandidateCount::mutate(|c| *c += 1);
+		}
+
+		/// Set the desired member count; if lower than the current count, then seats will not be up
+		/// election when they expire. If more, then a new vote will be started if one is not
+		/// already in progress.
+		#[weight = SimpleDispatchInfo::FixedOperational(10_000)]
```

Comment:
does one storage write and it's 10k. the one above does 3 mutations and costs 100x as much. wtf?

Reply:
about your 3 comments on the steepness of weights:

1. they are taken by approximately halving the weights that are decided for the other election (maybe a wrong decision, but that's not the main reason of your complaint) via benchmarking, which is _not through manual analysis_ but rather execution time measuring and _some manual juice added by me_. Some examples:

    1. submitting a candidate (as you recently discussed in some chats) is quite expensive with regards to computing the outcome of the election so I bumped it up a bit.
    2. removing a voter and adding one is more or less the same in terms of cpu/io but I manually reduced the weight of removing a bit for the same reason: it will potentially reduce the block time of the next election.


2. and more importantly this type of weight that you are analyzing, while I agree with it, is the `state-weight` which we, _could_, but at the moment have not yet introduced anywhere. Hence, I didn't apply it here either. The current weight numbers are merely a representation of computation time. Based on this, setting the desired seats could be arguably substantially less computationally intensive than submitting candidacy, hence the 100x difference.

Reply:
not sure i understand point (2) could you rephrase?

in general, weight should not be used to encode indirect costs because it is used to determine raw computation complexity of a block and is only incidentally used for fees.

if it's one-off indirect cost, then an additional non-weight fee is reasonable, to be charged in a discrete instance as part of the call. if it's an on-going cost, then an accordingly large deposit should be held (`ReservableCurrrency` can be used for this).

Reply:
> not sure i understand point (2) could you rephrase?

What I mean is that judging these based on how many mutate/read/writes they have is _correct_ but not something that we yet do; the current weight in all modules is merely based on execution time. regardless of what they do internally. Totally black-boxed. analyzing read/write/mutate etc. will be one of the early next steps after Kusama.  

> ... perverse incentive for validators to do less work by choosing transactions like these that are over-priced because they're accounting for a future cost, to be born by Some Other Guy.

Noted and I can conclude what the little bits of my manual tweaks are wrong. Will reverse them + fix the rest of them in further weight updates.